internal/checkcontroller/reconcile.go

```diff
@@ -0,0 +1,518 @@
+func (r *CheckControllerReconciler) MarkNodeUnschedulable(ctx context.Context, node *corev1.Node) error {
+	logger := log.FromContext(ctx).WithName("MarkNodeUnschedulable").WithValues("nodeName", node.Name)
+	logger.V(1).Info("Marking node as unschedulable")
+
+	node.Spec.Unschedulable = true
+	if err := r.Update(ctx, node); err != nil {
+		return fmt.Errorf("failed to update node %s to unschedulable: %w", node.Name, err)
+	}
+	return nil
+}
+
+// EvictPodsFromNode lists and evicts all non-critical pods from the specified node.
+func (r *CheckControllerReconciler) EvictPodsFromNode(ctx context.Context, nodeName string) error {
+	logger := log.FromContext(ctx).WithName("EvictPodsFromNode").WithValues("nodeName", nodeName)
+	logger.V(1).Info("Listing pods on node")
+
+	podList := &corev1.PodList{}
+	if err := r.List(ctx, podList, client.MatchingFields{"spec.nodeName": nodeName}); err != nil {
+		return fmt.Errorf("failed to list pods on node %s: %w", nodeName, err)
+	}
+
+	for _, pod := range podList.Items {
+		if ShouldSkipEviction(pod) {
+			logger.V(1).Info("Skipping eviction for critical pod", "podName", pod.Name, "priority", pod.Spec.Priority)
+			continue
+		}
+
+		if err := r.EvictPod(ctx, &pod); err != nil {
+			return fmt.Errorf("failed to evict pod %s: %w", pod.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// ShouldSkipEviction determines if a pod should be skipped during eviction.
+// Pods with a priority of higher than 1000000 are considered critical and should not be evicted.
+func ShouldSkipEviction(pod corev1.Pod) bool {
+	return pod.Spec.Priority != nil && *pod.Spec.Priority > 1000000
+}
+
+// evictPod evicts the given pod using the Kubernetes eviction API.
+func (r *CheckControllerReconciler) EvictPod(ctx context.Context, pod *corev1.Pod) error {
+	logger := log.FromContext(ctx).WithName("evictPod")
+	logger.V(1).Info("Evicting pod", "podName", pod.Name, "namespace", pod.Namespace)
+
+	eviction := &policyv1.Eviction{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pod.Name,
+			Namespace: pod.Namespace,
+		},
+	}
+
+	if err := r.Client.SubResource("eviction").Create(ctx, pod, eviction); err != nil {
+		if apierrors.IsTooManyRequests(err) {
+			logger.V(1).Info("Pod eviction is rate-limited, retrying", "podName", pod.Name)
+			return nil
+		}
+		logger.Error(err, "Failed to evict pod", "podName", pod.Name)
+		return err
+	}
+
+	logger.V(1).Info("Successfully evicted pod", "podName", pod.Name)
+	return nil
+}
+
+// checkNodeCondition checks if the node with the given name has a custom condition set.
+func (r *CheckControllerReconciler) CheckNodeCondition(
+	ctx context.Context, node *corev1.Node, nodeConditionType corev1.NodeConditionType, conditionStatus corev1.ConditionStatus,
+) (bool, error) {
+
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == nodeConditionType && condition.Status == conditionStatus {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
```

Comment:
```suggestion
func (r *CheckControllerReconciler) CheckNodeCondition(
	ctx context.Context, node *corev1.Node, nodeConditionType corev1.NodeConditionType, conditionStatus corev1.ConditionStatus,
) bool {

	for _, condition := range node.Status.Conditions {
		if condition.Type == nodeConditionType && condition.Status == conditionStatus {
			return true
		}
	}

	return false
}
```
probably we don't need an error here?